# EARLY DETECTION OF LOAN DEFAULT RISK FOR SMES & YOUTH ENTREPRENEURS

## 1. Business Understanding

### 1.1 Problem Context

Small and medium enterprises (SMEs) and youth entrepreneurs in emerging markets face limited access to affordable credit due to high perceived default risk. Manual credit assessments are slow, subjective, and difficult to scale.

This project develops a **supervised machine learning model** to **predict early loan default risk**, enabling lenders to:

    Reduce non-performing loans
    Intervene early with high-risk borrowers
    Promote financial inclusion through fairer credit decisions

### 1.2 Stakeholders
   **Primary**: Credit officers, risk analysts, loan managers

   **Secondary**: SME lenders, fintech product teams, regulators

### 1.3 Objective
Can we predict whether a loan will default using borrower demographics, loan terms, and early repayment behavior?

    Target variable: default_flag (0 = non-default, 1 = default)

    Prediction type: Binary classification/Multi_Class

    Prediction timing: Early lifecycle

## 2. Environment Setup

In [1]:
#core library
import pandas as pd
import numpy as np

#visualization 
import seaborn as sns
import matplotlib.pyplot as plt


#Preprocessing
from sklearn.model_selection import train_test_split

#others to be imported as needed

## 3. Data Understanding

### 3.1  Data Overview

The dataset used in this analysis consists of **32,296 loan records** collected from **Micromart Africa’s digital lending platform** for the purpose of **loan default prediction**. It contains several key feature groups, including: **borrower attributes** (e.g., age, gender, and account age), **loan characteristics** (e.g., principal, interest, total loan amount, loan term, product type, and number of installments), and **repayment behavior indicators** (e.g., total amount paid, unpaid balance, days late, and numbers of late and on-time payments).

The dataset is structured as a **binary classification problem**, where the target variable indicates whether a borrower defaulted on a loan. It combines demographic, transactional, and behavioral information, enabling the development of supervised machine learning models for credit risk assessment.


### 3.2  Data Description

### 3.2.1  Importing the dataset

In [7]:
#importing and printing head
data = pd.read_csv("data/loans_dataset.csv")
data.head()

,BorrowerId,Gender,Age,AccountAgeDays,EntityAgent,EntityUnit,LoanId,Principal,Interest,LoanAmount,...,ProductId,is_default,NumInstallments,TotalAmountPaid,TotalUnpaidAmount,AvgInstallmentPaid,TotalOverpaid,AvgDaysLate,NumLatePayments,NumOnTimePayments
0,17739,2.0,49.0,417.0,44150,11,57110,6000.0,1500.0,7500.0,...,30246.0,1,4,7500.0,0.0,1875.000000,2750.0,12.0,4,0
1,16694,2.0,44.0,444.0,44175,51,53097,15000.0,3750.0,18750.0,...,30280.0,0,1,18750.0,0.0,18750.000000,NaN,-1.0,0,1
2,19392,1.0,36.0,381.0,44258,106,70268,8000.0,2000.0,10000.0,...,30246.0,1,0,NaN,NaN,NaN,NaN,NaN,0,0
3,24897,1.0,42.0,206.0,44306,154,72095,6000.0,1500.0,7500.0,...,30246.0,1,0,NaN,NaN,NaN,NaN,NaN,0,0
4,12064,2.0,38.0,567.0,44111,34,55419,6000.0,1800.0,7800.0,...,30264.0,1,14,5900.0,1900.0,536.363636,2457.3,155.0,11,0


### 3.2.2  Basic Structure

In [8]:
#Shape
data.shape

(32296, 26)

In [9]:
#data columns
data.columns

Index(['BorrowerId', 'Gender', 'Age', 'AccountAgeDays', 'EntityAgent',
       'EntityUnit', 'LoanId', 'Principal', 'Interest', 'LoanAmount',
       'LoanBalance', 'LoanTermDays', 'NumRollovers', 'ChannelUsed',
       'ApprovalStage', 'IsLoanNew', 'ProductId', 'is_default',
       'NumInstallments', 'TotalAmountPaid', 'TotalUnpaidAmount',
       'AvgInstallmentPaid', 'TotalOverpaid', 'AvgDaysLate', 'NumLatePayments',
       'NumOnTimePayments'],
      dtype='object')

### 3.2.3 Overview of column types and non-null values

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32296 entries, 0 to 32295
Data columns (total 26 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   BorrowerId          32296 non-null  int64  
 1   Gender              30930 non-null  float64
 2   Age                 30128 non-null  float64
 3   AccountAgeDays      31910 non-null  float64
 4   EntityAgent         32296 non-null  int64  
 5   EntityUnit          32296 non-null  int64  
 6   LoanId              32296 non-null  int64  
 7   Principal           32124 non-null  float64
 8   Interest            32296 non-null  float64
 9   LoanAmount          32286 non-null  float64
 10  LoanBalance         32286 non-null  float64
 11  LoanTermDays        32083 non-null  float64
 12  NumRollovers        32296 non-null  int64  
 13  ChannelUsed         26414 non-null  float64
 14  ApprovalStage       32283 non-null  float64
 15  IsLoanNew           32296 non-null  int64  
 16  Prod

### 3.2.4 Summary statistics

In [14]:
data.describe()

,BorrowerId,Gender,Age,AccountAgeDays,EntityAgent,EntityUnit,LoanId,Principal,Interest,LoanAmount,...,ProductId,is_default,NumInstallments,TotalAmountPaid,TotalUnpaidAmount,AvgInstallmentPaid,TotalOverpaid,AvgDaysLate,NumLatePayments,NumOnTimePayments
count,32296.000000,30930.000000,30128.000000,31910.000000,32296.000000,32296.000000,32296.000000,32124.000000,32296.000000,32286.000000,...,32286.000000,32296.000000,32296.000000,27900.000000,27900.000000,27900.000000,2.490800e+04,27411.000000,32296.000000,32296.000000
mean,18510.288859,1.658390,38.129149,413.263052,43839.997306,70.991361,59985.212596,14672.014643,5497.164271,20368.178932,...,29747.130738,0.588060,8.518021,17796.798576,2573.345328,4917.968760,1.067908e+04,12.514830,3.193213,2.222411
std,6763.583513,0.474257,11.242072,219.910408,4655.085648,56.628759,10795.718936,18725.578741,9677.731327,27751.683610,...,3913.098058,0.492192,23.132524,25654.406732,11324.088718,11997.830776,3.935656e+04,80.285632,4.989676,4.216664
min,2.000000,1.000000,-3.000000,63.000000,35.000000,3.000000,2.000000,100.000000,0.000000,0.000000,...,1.000000,0.000000,0.000000,0.000000,-313725.000000,0.000000,0.000000e+00,-97.000000,0.000000,0.000000
25%,13656.000000,1.000000,30.000000,252.000000,44133.000000,19.000000,51970.750000,5000.000000,1250.000000,6250.000000,...,30246.000000,0.000000,4.000000,6250.000000,0.000000,1562.500000,1.050000e+02,0.000000,0.000000,0.000000
50%,18284.500000,2.000000,38.000000,401.000000,44250.000000,45.000000,60044.500000,8000.000000,2000.000000,10000.000000,...,30246.000000,1.000000,4.000000,9618.500000,0.000000,2316.666666,1.680000e+03,4.000000,2.000000,1.000000
75%,23395.000000,2.000000,45.000000,534.000000,44351.000000,124.000000,68118.250000,15000.000000,3750.000000,18750.000000,...,30271.000000,1.000000,6.000000,16680.000000,0.000000,3750.000000,5.549707e+03,11.000000,4.000000,3.000000
max,33784.000000,2.000000,87.000000,1479.000000,45535.000000,166.000000,79135.000000,400000.000000,189000.000000,460000.000000,...,31355.000000,1.000000,365.000000,590000.000000,200000.000000,590000.000000,1.092408e+06,9116.000000,65.000000,83.000000


### 3.2.5  Missing values

In [15]:
data.isna().mean()*100

BorrowerId             0.000000
Gender                 4.229626
Age                    6.712906
AccountAgeDays         1.195194
EntityAgent            0.000000
EntityUnit             0.000000
LoanId                 0.000000
Principal              0.532574
Interest               0.000000
LoanAmount             0.030964
LoanBalance            0.030964
LoanTermDays           0.659524
NumRollovers           0.000000
ChannelUsed           18.212782
ApprovalStage          0.040253
IsLoanNew              0.000000
ProductId              0.030964
is_default             0.000000
NumInstallments        0.000000
TotalAmountPaid       13.611593
TotalUnpaidAmount     13.611593
AvgInstallmentPaid    13.611593
TotalOverpaid         22.875898
AvgDaysLate           15.125712
NumLatePayments        0.000000
NumOnTimePayments      0.000000
dtype: float64

### 3.2.6  Duplicates

In [16]:
data.duplicated().sum()

0

### 3.2.7  Data Summary

**Dataset Overview**
- **Records:** 32,296 loan observations  
- **Features:** 26 variables  
- **Scope:** Borrower demographics, loan characteristics, and repayment behavior  

**Structure**
- **Rows × Columns:** 32,296 × 26  
- **Data types:** 16 float, 10 integer  
- **Duplicates:** None detected  

**Data Quality**
- Core identifiers and loan fields are fully populated, indicating a well-maintained transactional dataset.
- **Demographics:**  
  - Gender missing: **4.2%**  
  - Age missing: **6.7%**
- **Repayment variables:**  
  - ChannelUsed missing: **18.2%**  
  - Post-disbursement metrics (`TotalAmountPaid`, `AvgInstallmentPaid`, `AvgDaysLate`): **~13–15%** missing  
  - TotalOverpaid: **22.9%** missing (expected due to rarity of overpayment)

**Target Variable**
- **is_default:** 0% missing (fully reliable for supervised learning)

**Suitability**
- Well-structured and clean dataset suitable for **credit risk modeling** and **loan default prediction**, with rich demographic, transactional, and behavioral signals.


## 4. Data Preparation

In [ ]:
### 4.1   Data Cleaning

In [ ]:
### 4.1.1  Making a copy of the data

In [ ]:
### 4.1.2  Removing Duplicates

In [ ]:
### 4.1.3  Handling missing values and standardizing columns

## 5. Exploratory Data Analysis (EDA)

### 5.1 Univariate Analysis

### 5.2 Bivariate Analysis

### 5.3 Multivariate Analysis

## 6: Feature Engineering

## 7: Baseline Models

## 8: Advanced Models